### PR TITLE
Fix Windows platform file access to allow file sharing with external programs

### LINKED ISSUE #28036: Pressing F5 sometimes opens project manager due to Windows file locking limitations
**Godot version:**
3.1 stable

**OS/device including version:**
Windows 10

**Issue description:**
Every so often, when I press F5 to start my project, the Project list opens instead.

### PR DESCRIPTION
**Pull request by mhilbrunner**
This restores Windows platform file handling back to open files non-exlusively by default, as was the case before October 2018. (See https://github.com/godotengine/godot/commit/b902a2f2a7438810cdcb053568ed5c27089b1e8a)

It seems back then it was unintentionally changed while fixing warnings.

This is fixed by changing the open call from `_wfopen_s()` to `_wfsopen()`, thus being able to set the file sharing constant to `_SH_DENYNO` while still preserving security features of `_wfopen_s()`, like parameter validation, meaning the warning stays fixed.

Fixes #28036.

I tested this change on Windows as extensively as I could and found no issues.

---

<details>
  <summary>Details</summary>
  
Back then, while fixing warnings for *MSVC*, the function used for opening files was changed from `_wfopen()` to `_wfopen_s()` as suggested by the warning `C4996`. (`"This function may be unsafe, consider using _wfopen_s instead."`)

This new function
1. did parameter validation and thus avoided some possible security issues due to nil pointers or wrongly terminated strings
2. it also changed the default file sharing for opened files from `_SH_DENYNO` (which was the implicit default for the previous `_wfopen()`) to `_SH_SECURE`.

`_SH_DENYNO` means every opened file could be opened by other calls (like is the default on other operating systems).
`_SH_SECURE` means if the file is opened with READ access, others can still read the same file (but not write to it), but if it is opened with WRITE access, others can't open it at all, not even to read.

As this file sharing access is implied as default, not a parameter for either function, it seems this change was neither intended nor noticed.

This led to rarely occuring bugs on Windows, i.e. due to random access by Antivirus processes, or Godot/Windows not closing a file handle fast enough while trying to open it again elsewhere (i.e. `project.godot`, instead showing the Project Manager, or saving shaders/debugging the game).

What this PR does is change the file access to a third method, `_wfsopen()`. This is still secure, doing parameter validation and thus avoids the warning, but it allows us to actually SET the file sharing parameter. And we set it to `_SH_DENYNO`, as it was implicitely before the change. (And as it currently is on all non-Windows platforms, where file sharing restrictions don't exist by default.)

Warning `C4996` should really have been pointing this out. It should've been `_wfsopen()` all along. Let's hope this banishes those annoying, rare errors for all eternity.

The only interesting lines are 116 and 316, changing the function used for opening in `FileAccessWindows::_open()` and `FileAccessWindows::file_exists()` respectively, everything else are formatting fixes.

Links:
[_wfopen() docs ](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/fopen-wfopen?view=msvc-160)
[_wfopen_s() docs](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/fopen-s-wfopen-s?view=msvc-160)
[_wfsopen() docs](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/fsopen-wfsopen?view=msvc-160)
</details>

### CODE REVIEW

**Comment by mhilbrunner** [comment]
Short explanation on sharing modes:

`_SH_DENYNO`: No restrictions. Multiple processes can open the file for reading, writing, or any combination thereof.
This was the default before 2018, *is* the default for vanilla `fopen()`, and is the case on all non-Windows platforms.
This PR changes all file access back to this.
This allows `tail -f` to read the file while it's being written to, but also allows e.g. multiple log writers to `append` to the same log file.
`_SH_DENYRW`: Denies other processes read and write access. Nothing allowed.
`_SH_DENYRD`: Denies other processes read access. They can still write.
`_SH_DENYWR`: Denies other processes write access. They can still read.
`_SH_SECURE`: The current default since 2018, before this PR. If the file is open for reading, other processes can read, but not write. If the file is open for writing, other processes can do nothing, not even read.

**Review by RandomShaper — APPROVED** [review]

**Comment by RandomShaper** [comment]
I'm not very fond of the presence of the cosmetic changes in the same commit, but looks good to me overall.

**Comment by mhilbrunner** [comment]
I have no problem seperating those to a second commit if you want :) Just didn't bother because it was only 9 lines.

File: drivers/windows/file_access_windows.cpp [hunk 1]
```diff
@@ -77,7 +77,7 @@ Error FileAccessWindows::_open(const String &p_path, int p_mode_flags) {
 		return ERR_INVALID_PARAMETER;
 	}
 
-	/* pretty much every implementation that uses fopen as primary
+	/* Pretty much every implementation that uses fopen as primary
```
Comment (akien-mga) [inline drivers/windows/file_access_windows.cpp:80]: If fixing comment style, it should be fixed fully ;) (end with a dot)

https://docs.godotengine.org/en/latest/community/contributing/code_style_guidelines.html#comment-style-guide

**Comment by akien-mga** [comment]
> I have no problem seperating those to a second commit if you want :) Just didn't bother because it was only 9 lines.

I think it's OK in this case as it's indeed just a few lines. For more substantial changes keeping them separate is good practice indeed.

**Comment by mhilbrunner** [comment]
Seperated cosmetic changes into a second commit, added a dot :)

**Comment by akien-mga** [comment]
Thanks!

**Comment by bruvzg** [comment]
MinGW (9.0.0, GCC 11.2.0) build on macOS fails with `error: '_SH_DENYNO' was not declared in this scope` due to missing `#include <share.h>` in the `file_access_windows.cpp`.

**Comment by akien-mga** [comment]
> MinGW (9.0.0, GCC 11.2.0) build on macOS fails with `error: '_SH_DENYNO' was not declared in this scope` due to missing `#include <share.h>` in the `file_access_windows.cpp`.

Fixed by #51470.

**Comment by akien-mga** [comment]
Cherry-picked for 3.4.

**Comment by akien-mga** [comment]
Cherry-picked for 3.3.3.

**Pull request merged**